File: tracing/tracing/value/histogram_unittest.py
# Copyright 2017 The Chromium Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

import math
import unittest

from tracing.value import histogram


class RangeUnittest(unittest.TestCase):
  def testAddValue(self):
    r = histogram.Range()
    self.assertEqual(r.empty, True)
    r.AddValue(1)
    self.assertEqual(r.empty, False)
    self.assertEqual(r.min, 1)
    self.assertEqual(r.max, 1)
    self.assertEqual(r.center, 1)
    r.AddValue(2)
    self.assertEqual(r.empty, False)
    self.assertEqual(r.min, 1)
    self.assertEqual(r.max, 2)
    self.assertEqual(r.center, 1.5)


class RunningStatisticsUnittest(unittest.TestCase):
  def _Run(self, data):
    running = histogram.RunningStatistics()
    for datum in data:
      running.Add(datum)
    return running

  def testStatistics(self):
    running = self._Run([1, 2, 3])
    self.assertEqual(running.sum, 6)
    self.assertEqual(running.mean, 2)
    self.assertEqual(running.min, 1)
    self.assertEqual(running.max, 3)
    self.assertEqual(running.variance, 1)
    self.assertEqual(running.stddev, 1)
    self.assertEqual(running.geometric_mean, math.pow(6, 1./3))
    self.assertEqual(running.count, 3)

    running = self._Run([2, 4, 4, 2])
    self.assertEqual(running.sum, 12)
    self.assertEqual(running.mean, 3)
    self.assertEqual(running.min, 2)
    self.assertEqual(running.max, 4)
    self.assertEqual(running.variance, 4./3)
    self.assertEqual(running.stddev, math.sqrt(4./3))
    self.assertAlmostEqual(running.geometric_mean, math.pow(64, 1./4))
    self.assertEqual(running.count, 4)

  def testMerge(self):
    def Compare(data1, data2):
      a_running = self._Run(data1 + data2)
      b_running = self._Run(data1).Merge(self._Run(data2))
      CompareRunningStatistics(a_running, b_running)
      a_running = histogram.RunningStatistics.FromDict(a_running.AsDict())
      CompareRunningStatistics(a_running, b_running)
      b_running = histogram.RunningStatistics.FromDict(b_running.AsDict())
      CompareRunningStatistics(a_running, b_running)

    def CompareRunningStatistics(a_running, b_running):
      self.assertEqual(a_running.sum, b_running.sum)
      self.assertEqual(a_running.mean, b_running.mean)
      self.assertEqual(a_running.min, b_running.min)
      self.assertEqual(a_running.max, b_running.max)
      self.assertAlmostEqual(a_running.variance, b_running.variance)
      self.assertAlmostEqual(a_running.stddev, b_running.stddev)
      self.assertAlmostEqual(a_running.geometric_mean, b_running.geometric_mean)
      self.assertEqual(a_running.count, b_running.count)

    Compare([], [])
    Compare([], [1, 2, 3])
    Compare([1, 2, 3], [])
    Compare([1, 2, 3], [10, 20, 100])
    Compare([1, 1, 1, 1, 1], [10, 20, 10, 40])

File: tracing/tracing/value/histogram_unittest.cc
// Copyright 2019 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "tracing/tracing/value/histogram.h"

#include <cmath>
#include <sstream>

#include "testing/gtest/include/gtest/gtest.h"
#include "tracing/tracing/proto/histogram.pb.h"

namespace catapult {

namespace proto = tracing::tracing::proto;

proto::UnitAndDirection UnitWhatever() {
  proto::UnitAndDirection unit;
  unit.set_unit(proto::UNITLESS);
  unit.set_improvement_direction(proto::NOT_SPECIFIED);
  return unit;
}

TEST(HistogramTest, WritesCorrectNameToProto) {
  HistogramBuilder builder("my name", UnitWhatever());

  auto histogram = builder.toProto();

  EXPECT_EQ(histogram->name(), "my name");
}

TEST(HistogramTest, WritesCorrectDescriptionToProto) {
  HistogramBuilder builder("", UnitWhatever());

  builder.set_description("desc!");

  auto histogram = builder.toProto();

  EXPECT_EQ(histogram->description(), "desc!");
}

TEST(HistogramTest, WritesCorrectUnitToProto) {
  proto::UnitAndDirection unit;
  unit.set_unit(proto::TS_MS);
  unit.set_improvement_direction(proto::BIGGER_IS_BETTER);
  HistogramBuilder builder("", unit);

  auto histogram = builder.toProto();

  EXPECT_EQ(histogram->unit().unit(), proto::TS_MS);
  EXPECT_EQ(histogram->unit().improvement_direction(), proto::BIGGER_IS_BETTER);
}

TEST(HistogramTest, WritesSmallNumberOfSamplesToProtoInOrder) {
  HistogramBuilder builder("", UnitWhatever());

  builder.AddSample(1.0);
  builder.AddSample(2.0);
  builder.AddSample(3.0);
  builder.AddSample(4.0);

  auto histogram = builder.toProto();

  EXPECT_EQ(histogram->sample_values_size(), 4);
  EXPECT_EQ(histogram->sample_values(0), 1.0);
  EXPECT_EQ(histogram->sample_values(1), 2.0);
  EXPECT_EQ(histogram->sample_values(2), 3.0);
  EXPECT_EQ(histogram->sample_values(3), 4.0);
}

TEST(HistogramTest, StartsUniformlySamplingAfterReachingMaxNumSamples) {
  HistogramBuilder builder("", UnitWhatever());

  for (int i = 0; i < 100; ++i) {
    builder.AddSample(i);
  }

  auto histogram = builder.toProto();

  EXPECT_EQ(histogram->sample_values_size(), 10)
      << "Did not expect num samples to grow beyond 10, which is the default "
      << "max number of samples";

  // Values will get randomly thrown away, but do some spot checks that they
  // at least are in range.
  ASSERT_LE(0, histogram->sample_values(0));
  ASSERT_GE(99, histogram->sample_values(0));
  ASSERT_LE(0, histogram->sample_values(1));
  ASSERT_GE(99, histogram->sample_values(1));
  ASSERT_LE(0, histogram->sample_values(9));
  ASSERT_GE(99, histogram->sample_values(9));
}

TEST(HistogramTest, WritesCorrectRunningStatisticsToProto) {
  HistogramBuilder builder("", UnitWhatever());

  builder.AddSample(10.0);
  builder.AddSample(20.0);
  builder.AddSample(30.0);
  builder.AddSample(40.0);

  auto histogram = builder.toProto();

  EXPECT_EQ(histogram->running().count(), 4);
  EXPECT_EQ(histogram->running().max(), 40.0);
  EXPECT_FLOAT_EQ(histogram->running().meanlogs(), 3.0970986);
  EXPECT_EQ(histogram->running().mean(), 25.0);
  EXPECT_EQ(histogram->running().min(), 10.0);
  EXPECT_EQ(histogram->running().sum(), 100.0);
  EXPECT_FLOAT_EQ(histogram->running().variance(), 166.6667);
}

TEST(HistogramTest, DoesNotWriteMeanlogsIfNegativeSampleAdded) {
  HistogramBuilder builder("", UnitWhatever());

  builder.AddSample(20.0);
  builder.AddSample(-1.0);

  auto histogram = builder.toProto();

  ASSERT_EQ(histogram->running().meanlogs(), 0);
}

}  // namespace catapult

